Client/Frontend/Brave Today/Composer/FeedDataSource.swift

```diff
@@ -0,0 +1,544 @@
+                try filePaths.forEach {
+                    var fileUrl = braveTodayPath
+                    fileUrl.appendPathComponent($0)
+                    try fileManager.removeItem(atPath: fileUrl.path)
+                }
+            }
+        } catch {
+            logger.error("Could not remove cached files")
+            return false
+        }
+        return true
+    }
+    
+    // MARK: - Sources
+    
+    /// Get a map of customized sources IDs and their overridden enabled states
+    var customizedSources: [String: Bool] {
+        let all = BraveTodaySourceMO.all()
+        return all.reduce(into: [:]) { (result, source) in
+            result[source.publisherID] = source.enabled
+        }
+    }
+    
+    /// Whether or not a source is currently enabled (whether or not by default or by a user changing
+    /// said default)
+    func isSourceEnabled(_ source: FeedItem.Source) -> Bool {
+        BraveTodaySourceMO.get(fromId: source.id)?.enabled ?? source.isDefault
+    }
+    
+    /// Toggle a source's enabled status
+    func toggleSource(_ source: FeedItem.Source, enabled: Bool) {
+        BraveTodaySourceMO.setEnabled(forId: source.id, enabled: enabled)
+        
+        if let cards = state.cards, cards.isEmpty && enabled {
+            // If we're enabling a source and we don't have any items because their source selection was
+            // causing an empty generation, regenerate the cards
+            self.reloadCards(from: self.items, sources: self.sources)
+        }
+    }
+    
+    /// Toggle an entire category on or off
+    func toggleCategory(_ category: String, enabled: Bool) {
+        let sourcesInCategory = sources.filter { $0.category == category }
+        if sourcesInCategory.isEmpty {
+            return
+        }
+        BraveTodaySourceMO.setEnabled(forIds: sourcesInCategory.map(\.id), enabled: enabled)
+        
+        if let cards = state.cards, cards.isEmpty && enabled {
+            // If we're enabling a category and we don't have any items because their source selection was
+            // causing an empty generation, regenerate the cards
+            self.reloadCards(from: self.items, sources: self.sources)
+        }
+    }
+    
+    /// Reset all source settings back to default
+    func resetSourcesToDefault() {
+        BraveTodaySourceMO.resetSourceSelection()
+    }
+    
+    // MARK: - Card Generation
+    
+    /// Scores and generates cards from a set of items and sources
+    private func reloadCards(
+        from items: [FeedItem.Content],
+        sources: [FeedItem.Source],
+        completion: (() -> Void)? = nil
+    ) {
+        // Only allow 1 reload at a time
+        // Since the scoring/generation work hops between threads a lot it would be possible for reloads to
+        // mess up the ending state or update the cards
+        reloadQueue.async { [weak self] in
+            let group = DispatchGroup()
```

Comment:
We use the DispatchGroup because the contents of the block are not serial, therefore `reloadQueue` would finish executing immediately and not really work as a serial queue